### PR TITLE
05 building portfolio with transactions

### PR DESCRIPTION
**Pull request by fedefailla18**

### CODE REVIEW

**Comment by facum15** [comment]
Deadline is approaching...

File: src/integration-test/groovy/com/importer/fileimporter/controller/WebControllerIntegrationSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import com.importer.fileimporter.service.ProcessFile
+import org.springframework.boot.test.mock.mockito.MockBean
+import org.springframework.core.io.ClassPathResource
+import org.springframework.http.MediaType
+import org.springframework.mock.web.MockMultipartFile
+import org.springframework.security.test.context.support.WithMockUser
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
+import spock.lang.Ignore
+
+class WebControllerIntegrationSpec extends BaseIntegrationSpec {
```
Comment (fedefailla18) [inline src/integration-test/groovy/com/importer/fileimporter/controller/WebControllerIntegrationSpec.groovy:18]: this doesn't seem to be an Integration test, move it to the right path

File: src/main/java/com/importer/fileimporter/service/CoinInformationService.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Service
+@Slf4j
+public class CoinInformationService {
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/service/CoinInformationService.java:19]: this looks more like a helper than a service layer

File: src/main/java/com/importer/fileimporter/service/CryptoCompareService.java [hunk 3]
```diff
@@ -2,38 +2,52 @@
...
 import org.springframework.web.reactive.function.client.WebClient;
 import reactor.core.publisher.Mono;
 
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 @Service
+@Slf4j
 public class CryptoCompareService {
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/service/CryptoCompareService.java:18]: isn't this a proxy?

File: src/main/java/com/importer/fileimporter/service/GetSymbolHistoricPriceService.java [hunk 4]
```diff
@@ -2,55 +2,118 @@
...
 import java.time.format.DateTimeFormatter;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 
 @RequiredArgsConstructor
 @Service
 @Slf4j
 public class GetSymbolHistoricPriceService {
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/service/GetSymbolHistoricPriceService.java:23]: this looks more like a facade

File: src/main/java/com/importer/fileimporter/utils/ProcessFileServiceUtils.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+@UtilityClass
+public class ProcessFileServiceUtils {
+
+    public final String PAIR_KEY = "Pair";
+    public final String EXECUTED_KEY = "Executed";
+    public final String AMOUNT_KEY = "Amount";
+    public final String FEE_KEY = "Fee";
+    public final String PRICE_KEY = "Price";
+    public final String DATE_KEY = "Date(UTC)";
+    public final String SIDE_KEY = "Side";
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/utils/ProcessFileServiceUtils.java:20]: we should have a class mapping this keys to make it dynamically and be able to do it for multiple files with different names

**Review by fedefailla18 — COMMENTED** [review]
this looks like it has https://github.com/fedefailla18/file-importer/pull/4 and https://github.com/fedefailla18/file-importer/pull/5 PRs here so we should be fine merging this to master and closing the other PR if needed

File: src/integration-test/groovy/com/importer/fileimporter/DatabaseConnectionSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+
+import org.postgresql.Driver
+import org.springframework.jdbc.datasource.SimpleDriverDataSource
+
+import java.sql.Connection
+import java.sql.DriverManager
+import java.sql.SQLException
+
+class DatabaseConnectionSpec extends BaseIntegrationSpec {
```
Comment (fedefailla18) [inline src/integration-test/groovy/com/importer/fileimporter/DatabaseConnectionSpec.groovy:11]: this should be named DatabaseConnectionIntegrationSpec

File: src/integration-test/groovy/com/importer/fileimporter/service/CryptoCompareServiceSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package com.importer.fileimporter.service
+
+import com.importer.fileimporter.BaseIntegrationSpec
+import com.importer.fileimporter.config.integration.CryptoCompareConfig
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.web.reactive.function.client.WebClient
+
+class CryptoCompareServiceSpec extends BaseIntegrationSpec {
```
Comment (fedefailla18) [inline src/integration-test/groovy/com/importer/fileimporter/service/CryptoCompareServiceSpec.groovy:8]: should be named CryptoCompareServiceIntegrationSpec

File: src/main/java/com/importer/fileimporter/config/security/WebSecurityConfig.java [hunk 8]
```diff
@@ -29,12 +29,12 @@ public class WebSecurityConfig {
...
     private final AuthEntryPointJwt unauthorizedHandler;
 
-    @Bean
+//    @Bean
     public AuthTokenFilter authenticationJwtTokenFilter() {
         return new AuthTokenFilter();
     }
 
-    @Bean
+//    @Bean
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/config/security/WebSecurityConfig.java:37]: is this comment necessary?

File: src/main/java/com/importer/fileimporter/config/security/WebSecurityConfig.java [hunk 9]
```diff
@@ -29,12 +29,12 @@ public class WebSecurityConfig {
 
     private final AuthEntryPointJwt unauthorizedHandler;
 
-    @Bean
+//    @Bean
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/config/security/WebSecurityConfig.java:32]: is this comment necessary?

File: src/main/java/com/importer/fileimporter/dto/integration/CryptoCompareSinglePricingResponse.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.importer.fileimporter.dto.integration;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+public class CryptoCompareSinglePricingResponse {
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/dto/integration/CryptoCompareSinglePricingResponse.java:8]: ios this used at all?

File: src/main/java/com/importer/fileimporter/entity/TransactionId.java [hunk 11]
```diff
@@ -28,10 +28,10 @@ public class TransactionId implements Serializable {
     @Column(name = "pair", nullable = false, length = 12)
     private String pair;
 
-    @Column(name = "price", nullable = false, precision = 12, scale = 5)
+//    @Column(name = "price", nullable = false, precision = 13, scale = 13)
     private BigDecimal price;
 
-    @Column(name = "executed", nullable = false, precision = 20, scale = 5)
+//    @Column(name = "executed", nullable = false, precision = 20, scale = 20)
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/entity/TransactionId.java:34]: can we delete this comments?

File: src/main/java/com/importer/fileimporter/service/usecase/CalculateAmountSpent.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+@RequiredArgsConstructor
+@Service
+public class CalculateAmountSpent {
```
Comment (fedefailla18) [inline src/main/java/com/importer/fileimporter/service/usecase/CalculateAmountSpent.java:21]: Create test

File: src/main/java/com/importer/fileimporter/service/GetSymbolHistoricPriceService.java (hunk 4 above)
Comment (fedefailla18) [reply src/main/java/com/importer/fileimporter/service/GetSymbolHistoricPriceService.java:23]: renamed to Helper

File: src/integration-test/groovy/com/importer/fileimporter/DatabaseConnectionSpec.groovy (hunk 6 above)
Comment (fedefailla18) [reply src/integration-test/groovy/com/importer/fileimporter/DatabaseConnectionSpec.groovy:11]: I deleted it

**Comment by fedefailla18** [comment]
Addressed some comments and adding more tests. Still we need to confirm that the whole integration works (calculate amount spent and avgEntryPrice. CoinInformationHelper should hold all this methods

**Review by fedefailla18 — COMMENTED** [review]
These changes add valuable references to the codebase, allowing us to gain insights into our transaction history data by symbol. The introduction of integration tests as their own module, along with several unit tests (with only two currently failing), enhances our testing framework significantly.

Overall, the branch is in good shape and provides a solid foundation for future development. Therefore, I'm merging this branch. Any additional changes should be made in separate branches taken from the main branch.

**Pull request merged**